src/lib/message/ExchangeContext.cpp

```diff
@@ -0,0 +1,1664 @@
+ *
+ *  @param[in]    sendFlags          Flags set by the application for the CHIP message being sent.
+ *
+ *
+ *  @retval  #CHIP_ERROR_BUFFER_TOO_SMALL  If the message buffer does not have sufficient space
+ *                                          for encoding the exchange header.
+ *  @retval  #CHIP_NO_ERROR                If encoding of the message was successful.
+ */
+CHIP_ERROR ExchangeContext::EncodeExchHeader(ChipExchangeHeader * exchangeHeader, uint32_t profileId, uint8_t msgType,
+                                             PacketBuffer * msgBuf, uint16_t sendFlags)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    // Fill the exchange header to the message buffer.
+    exchangeHeader->Version     = kChipExchangeVersion_V1;
+    exchangeHeader->ExchangeId  = ExchangeId;
+    exchangeHeader->ProfileId   = profileId;
+    exchangeHeader->MessageType = msgType;
+    // sendFlags under special circumstances (such as a retransmission
+    // of the remote alarm) can override the initiator flag in the
+    // exchange header.  The semantics here really is: use the
+    // ExchangeId in the namespace of the SourceNodeId
+    exchangeHeader->Flags = (IsInitiator() || (sendFlags & kSendFlag_FromInitiator)) ? kChipExchangeFlag_Initiator : 0;
+
+    // RMP PreProcess Checks and Flag setting
+    if (mMsgProtocolVersion == kChipMessageVersion_V2)
+    {
+#if CHIP_CONFIG_ENABLE_RELIABLE_MESSAGING
+        // If there is a pending acknowledgment piggyback it on this message.
+        // If there is no pending acknowledgment piggyback the last Ack that was sent.
+        //   - HasPeerRequestedAck() is used to verify that AckId field is valid
+        //     to avoid piggybacking uninitialized AckId.
+        if (HasPeerRequestedAck())
+        {
+            // Expire any virtual ticks that have expired so all wakeup sources reflect the current time
+            ExchangeMgr->RMPExpireTicks();
+
+            exchangeHeader->Flags |= kChipExchangeFlag_AckId;
+            exchangeHeader->AckMsgId = mPendingPeerAckId;
+
+            // Set AckPending flag to false after setting the Ack flag;
+            SetAckPending(false);
+
+            // Schedule next physical wakeup
+            ExchangeMgr->RMPStartTimer();
+
+#if defined(DEBUG)
+            ChipLogProgress(ExchangeManager, "Piggybacking Ack for MsgId:%08" PRIX32 " with msg", mPendingPeerAckId);
+#endif
+        }
+
+        // Assert the flag if message requires an Ack back;
+        if ((sendFlags & kSendFlag_RequestAck) && !IsRMPControlMessage(profileId, msgType))
+        {
+            exchangeHeader->Flags |= kChipExchangeFlag_NeedsAck;
+        }
+#endif // CHIP_CONFIG_ENABLE_RELIABLE_MESSAGING
+    }
+
+    err = ChipExchangeManager::PrependHeader(exchangeHeader, msgBuf);
+
+    return err;
+}
+
+/**
+ *  Cancel the Trickle retransmission mechanism.
+ *
+ */
+void ExchangeContext::CancelRetrans()
+{
+    // NOTE: modify for other retransmission schemes
```

Comment:
Should there be an issue tracking something here? (Again, sorry for the vague nature of things, not quite sure the intent of the source code)